### ISSUE TITLE
Bug: `--ignore-pattern` doesn't work well with `--flag unstable_config_lookup_from_file`

### ISSUE THREAD

**Issue by mdjermanovic** [eslint/eslint#18948]
### Environment

Node version: v20.12.0
npm version: 10.5.0
Local ESLint version: `issue18385c` branch
Global ESLint version:
Operating System: windows

### What parser are you using?

Default (Espree)

### What did you do?

There's an `eslint.config.js` file in `tests/lib/eslint`, so when the files in this directory are linted with `--flag unstable_config_lookup_from_file`, that config file is used.

Command (from the project root):

```bash
$ node bin/eslint tests/lib/eslint --ignore-pattern tests/lib/eslint/legacy-eslint.js --flag unstable_config_lookup_from_file
```

### What did you expect to happen?

`tests/lib/eslint/legacy-eslint.js` to be ignored.

### What actually happened?

It isn't ignored. The output is:

```
C:\projects\eslint\tests\lib\eslint\eslint.js
     7:1   error    'use strict' is unnecessary inside of modules                                       strict
   131:34  warning  Unused eslint-disable directive (no problems were reported from 'no-invalid-this')
   202:70  warning  Unused eslint-disable directive (no problems were reported from 'no-new')
   324:17  warning  Unused eslint-disable directive (no problems were reported from 'no-new')
   993:55  error    Strings must use doublequote                                                        quotes
  5787:17  error    Definition for rule 'n/no-unsupported-features/node-builtins' was not found         n/no-unsupported-features/node-builtins

C:\projects\eslint\tests\lib\eslint\legacy-eslint.js
     7:1   error    'use strict' is unnecessary inside of modules                                            strict
   103:34  warning  Unused eslint-disable directive (no problems were reported from 'no-invalid-this')
   158:17  warning  Unused eslint-disable directive (no problems were reported from 'no-new')
   167:65  warning  Unused eslint-disable directive (no problems were reported from 'no-new')
   938:18  error    Strings must use doublequote                                                             quotes
   942:13  warning  Unused eslint-disable directive (no problems were reported from 'no-underscore-dangle')
  5149:25  warning  Unused eslint-disable directive (no problems were reported from 'no-new')
  5175:21  warning  Unused eslint-disable directive (no problems were reported from 'no-new')
  5304:21  warning  Unused eslint-disable directive (no problems were reported from 'no-new')

✖ 15 problems (5 errors, 10 warnings)
  4 errors and 10 warnings potentially fixable with the `--fix` option.
```

### Link to Minimal Reproducible Example

https://github.com/eslint/eslint/tree/issue18385c

### Participation

- [x] I am willing to submit a pull request for this issue.

### Additional comments

Related: https://github.com/eslint/eslint/pull/18742#discussion_r1770630540

**Comment by nzakas** [comment]
Confirmed. Go ahead and submit a PR when ready.

**Comment by mdjermanovic** [comment]
This doesn't seem quite easy to implement to always work correctly. For example, with this command:

```sh
npx eslint . --ignore-pattern {foo/bar,baz}/a*.js --flag unstable_config_lookup_from_file
```

if there's a config file in `foo/bar`, when creating a config array for that config file, the ignore pattern should be transformed to `a*.js`, I believe. But that requires parsing and analyzing the pattern, which seems fairly complex.

A solution that wouldn't require analyzing patterns could be to introduce a `basePath` property on config objects and update `@eslint/config-array` to treat patterns in a config object as relative to that base path, if present, instead of the config array's base path. That might also be helpful if we allow for configuration cascading (automatically merging config files, like eslintrc without `root: true`) in the future. For patterns passed through `--ignore-pattern`, `basePath` in the config object where we're storing them would be set to `cwd` because they're relative to the current working directory.

Thoughts?

**Comment by nzakas** [comment]
> A solution that wouldn't require analyzing patterns could be to introduce a `basePath` property on config objects and update `@eslint/config-array` to treat patterns in a config object as relative to that base path,

:+1: I was actually pondering if this would be useful over the break. It seems like it would make mixing and matching configs from different directories a lot easier for end users. We would need an RFC for this.

**Comment by fasttime** [comment]
> A solution that wouldn't require analyzing patterns could be to introduce a `basePath` property on config objects and update `@eslint/config-array` to treat patterns in a config object as relative to that base path, if present, instead of the config array's base path.

I think that adding a `basePath` property to a config (or more generally, overriding a config array's default base path) could also address the use case of matching external files when the user has no control over the current directory as in #19118. I'd be also in favor of this solution 👍 .

**Comment by mdjermanovic** [comment]
I'm not sure what the scope of the RFC should be, i.e., what use cases should be addressed.

My idea for this particular issue was that ESLint automatically sets `basePath` in config objects when needed (currently, to `cwd` for ignore patterns that are relative to the current working directory), not that users can specify `basePath` in config objects in their config files, so this is more an internal detail than new functionality.

**Comment by nzakas** [comment]
> My idea for this particular issue was that ESLint automatically sets `basePath` in config objects when needed (currently, to `cwd` for ignore patterns that are relative to the current working directory), not that users can specify `basePath` in config objects in their config files, so this is more an internal detail than new functionality.

Hmmm, okay, that's very different than what I had in mind. I'm not sure I understand how an internal-only `basePath` property would work in practice. Can you explain how you envision that working?

> I'm not sure what the scope of the RFC should be, i.e., what use cases should be addressed.

I also think that's why an end-user accessible `basePath` property could be helpful. It seems like there is a ready list of use cases for that type of functionality.

**Comment by mdjermanovic** [comment]
> > My idea for this particular issue was that ESLint automatically sets `basePath` in config objects when needed (currently, to `cwd` for ignore patterns that are relative to the current working directory), not that users can specify `basePath` in config objects in their config files, so this is more an internal detail than new functionality.
> 
> Hmmm, okay, that's very different than what I had in mind. I'm not sure I understand how an internal-only `basePath` property would work in practice. Can you explain how you envision that working?

Currently in `@eslint/config-array` matching is always as follows: from the passed file path, it calculates path relative to the config array's `basePath` and matches that relative path with `files`/`ignores` patterns.

The change would be to, when matching with `files`/`ignores` patterns from a config object that has its own `basePath` property, it would calculate path relative to the config object's `basePath` instead of config array's `basePath`. The value of config object's `basePath` would be an absolute path set by ESLint. Technically, `basePath` on config objects could be "hidden" (e.g., a Symbol property) as I didn't envision end users setting this property.

We could allow user-specified `basePath`s but I'm still not sure for which use cases it would be useful. And those user-specified `basePath`s would probably be relative?

**Comment by nzakas** [comment]
> Technically, `basePath` on config objects could be "hidden" (e.g., a Symbol property) as I didn't envision end users setting this property.

I think if we do that then users will end up finding it, which is why I think making it an official key makes more sense. 

> We could allow user-specified `basePath`s but I'm still not sure for which use cases it would be useful. And those user-specified `basePath`s would probably be relative?

There are a couple of cases where this would be useful:

1. When you want one `eslint.config.js` file to apply to multiple directories when one or more of the directories is not a descendant of the current directory. 
2. When you want to manually merge configuration file from multiple different directories. Instead of needing to interpolate the `files` and `ignores` fields, you could just add a `basePath` to each object that maps to the original file location. (Basically, the DefinitelyTyped use case where they want JSON config in each project with `eslint.config.js` in a root directory.)

**Comment by simplg** [comment]
It might be related but when using "unstable_config_lookup_from_file" flag on a monorepo on a Mac, I couldn't get it to work.
Our monorepo is arranged like this : 
```
package.json
src/
---packages/
------app-1/
---------package.json
---------eslint.config.mjs
------app-2/
---------package.json
---------eslint.config.mjs
------api/
---------package.json
---------eslint.config.mjs
```

The error given is that it cannot find any eslint.config.(js|mjs|cjs) file. Even though I'm running `pnpx eslint --flag unstable_config_lookup_from_file .`. After enabling the debug option, I see it trying to find an eslint config file for a `.DS_Store` file found inside the packages folder.

Since I don't need eslint to check `.DS_Store` files, I wanted to add them into an ignore-pattern, but it doesn't seem to work at all. Eslint keeps trying to search for config files for this `.DS_Store` file.
Exemple of ignore-pattern tried:
* `pnpx eslint --flag unstable_config_lookup_from_file . --ignore-pattern '.DS_Store'`
* `pnpx eslint --flag unstable_config_lookup_from_file . --ignore-pattern '**/*.DS_Store'`
* `pnpx eslint --flag unstable_config_lookup_from_file . --ignore-pattern '*.DS_Store'`

**Comment by fasttime** [comment]
@simplg It looks like you need to add a `eslint.config.mjs` file at the root of your monorepo. This is required regardless of the `unstable_config_lookup_from_file` flag even when there are config files in subdirectories.

The fact that ESLint keeps looking for config files in hidden MacOS directories is intended, because there could be config files there, but that shouldn't be a problem I think.

**Comment by simplg** [comment]
> [@simplg](https://github.com/simplg) It looks like you need to add a `eslint.config.mjs` file at the root of your monorepo. This is required regardless of the `unstable_config_lookup_from_file` flag even when there are config files in subdirectories.
> 
> The fact that ESLint keeps looking for config files in hidden MacOS directories is intended, because there could be config files there, but that shouldn't be a problem I think.

If I remove all `.DS_Store` and I use the exact same command (i.e. `npx eslint --flag unstable_config_lookup_from_file packages`), it's doing as intended (without looking for any eslint.config.mjs at the cwd) and does only search for eslint.config.mjs file for files that need to be linted. For example, since I only have files inside packages/app-1, packages/app-2, packages/app-3 and there is no more file inside packages, it only tries to look for eslint.config.mjs files inside those subdirectories. However as soon as a .DS_Store file is generated by Mac inside the packages directory, it breaks (which I would like to ignore with ignore-pattern but it doesn't seem to work).

According to the doc, with the flag enabled, eslint `Look up `eslint.config.js` from the file being linted`, so it does seem to be having the correct behaviour and doesn't seem to need an eslint.config.js file at the cwd if there is no file needing to be linted at cwd.

If needed, I can provide a working example for illustration sake.

**Comment by fasttime** [comment]
> [@eslint/eslint-tsc](https://github.com/orgs/eslint/teams/eslint-tsc) Let's answer this question now: shall we add a `basePath` to config objects? I'm 👍 as I think it solves this and other use cases.

There are important details to discuss but I'm in favor of this change.

**Comment by fasttime** [comment]
> However as soon as a .DS_Store file is generated by Mac inside the packages directory, it breaks

@simplg If that's a different problem than what we are discussing here, could you please [open a bug report](https://github.com/eslint/eslint/issues/new?template=bug-report.yml) to describe it? Thanks!

**Comment by sam3k** [comment]
We've agreed to accept this proposal in our 2025-02-06 TSC meeting and will move forward with an RFC. @mdjermanovic will work on the RFC.

**Comment by nzakas** [comment]
@mdjermanovic what's the status on this? Is https://github.com/eslint/rewrite/compare/main...mdjermanovic:rewrite:base-path ready for a PR?

**Comment by mdjermanovic** [comment]
I'll submit a PR soon. The code change is ready, I believe, but needs tests to complete a PR.
